javascript/ql/src/semmle/javascript/dataflow/TypeTracking.qll

```diff
@@ -16,17 +16,27 @@ private class PropertyName extends string {
     AccessPath::isAssignedInUniqueFile(this)
     or
     exists(AccessPath::getAnAssignmentTo(_, this))
+    or
+    this instanceof TypeTrackingPseudoProperty
   }
 }
 
+/**
+ * A pseudo-property that can be used in type-tracking.
+ */
+abstract class TypeTrackingPseudoProperty extends string {
+  bindingset[this]
+  TypeTrackingPseudoProperty() { any() }
+}
+
 private class OptionalPropertyName extends string {
   OptionalPropertyName() { this instanceof PropertyName or this = "" }
 }
 
 /**
  * A description of a step on an inter-procedural data flow path.
  */
-private newtype TStepSummary =
+newtype TStepSummary =
```

Comment:
Instead of making this public, could you move this to a file `dataflow/internal/StepSummary.qll` which is imported from here and in the promise library where it is needed?

Reply:
I moved both the type and the `StepSummary` class into the new `.qll` file, and this lead to some refactorization in a few places.